ducktape/tests/action.py

```diff
@@ -0,0 +1,230 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Provides pass/fail/aux-results reporting for specific test actions without
+# necessarily affecting the entire test outcome.
+#
+
+import traceback
+import inspect
+import time
+
+class TestAction (object):
+    # Accumulated test actions in various forms.
+    actions_root = list()
+    actions = list()
+    actions_by_fullname = dict()
+
+    def __init__ (self, name, instance=None, desc=None, parent=None, logger=None):
+        """
+            name     := Appended to parent's name to create unique action identifier
+            instance := Instance to test: allows multiple actions with the same name
+            desc     := Human readable description of this action
+            parent   := Parent action, either a TestAction object or a fully qualified
+                        action name (as returned by TestAction.fullname).
+            logger   := Logger to use. Only set this on your top-level action, it will be
+                        inherited to all sub-actions.
+        """
+        
+        super(TestAction, self).__init__()
+        self.name       = name
+        self.instance   = instance
+        self.desc       = desc
+        self.logger     = logger
+        self.start_time = time.time()
+        self.duration   = 999.999
+        # Resolve parent
+        if type(parent) == str:
+            if parent in actions:
+                parent = actions[parent]
+            else:
+                parent = None
+        self.parent   = parent
+
+        prefix=''
+        if self.parent:
+            # Add us to parent
+            self.parent.subs.append(self)
+            # Use parent's logger
+            if not self.logger and self.parent.logger:
+                self.logger = self.parent.logger
+            # Prefix our name with parent's name
+            prefix = self.parent.fullname + '.'
+
+
+        # If there is no description but the calling function has a doc string, use that,
+        # unless the parent action has the same one (a bit repetative).
+        # FIXME: Doesnt seem to work for class functions
+        #if not self.desc:
+        #    cfr   = inspect.currentframe().f_back
+        #    cname = cfr.f_code.co_name
+        #    doc   = cfr.f_globals[cname].__doc__
+        #    if not self.parent or doc != self.parent.desc:
+        #        self.desc = doc
+
+        if instance != None:
+            self.fullname = '%s%s@%s' % (prefix, name, instance)
+        else:
+            self.fullname = prefix + name
+
+        if not self.logger:
+            self.logger = self._dummy_logger
```

Comment:
Does this work? Don't we usually make `self.logger` a `logging.Logger` object, which seems incompatible with it being a bound method like `self._dummy_logger`?